File: src/main/java/iog/psg/bcc/jpi/ApiRequestExecutor.java
package iog.psg.bcc.jpi;

import java.util.concurrent.CompletionStage;

public interface ApiRequestExecutor {
    <T> CompletionStage<T> execute(iog.psg.bcc.BccApi.BccApiRequest<T> request) throws BccApiException;

}

File: src/main/scala/iog/psg/bcc/ApiRequestExecutor.scala
package iog.psg.bcc

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import iog.psg.bcc.BccApi.{BccApiRequest, BccApiResponse}

import scala.concurrent.{ExecutionContext, Future}

object ApiRequestExecutor extends ApiRequestExecutor

trait ApiRequestExecutor {

  def execute[T](request: BccApiRequest[T])(implicit ec: ExecutionContext, as: ActorSystem): Future[BccApiResponse[T]] =
    Http()
      .singleRequest(request.request)
      .flatMap(request.mapper)

}
